activesupport/lib/active_support/core_ext/object/json.rb

```diff
@@ -184,8 +188,10 @@ def as_json(options = nil) # :nodoc:
     end
 
     result = {}
```

Comment:
Should we just use `to_h` here?

```ruby
    if options
      subset.to_h { |k, v| [k.to_s, v.as_json(options.dup)] }
    else
      subset.to_h { |k, v| [k.to_s, v.as_json] }
    end
```

Reply:
Is it slower?

Reply:
It has a non-obvious (to me) time/memory trade off. It's about 15% (20% with YJIT) faster across all Hash sizes:

<details><summary>Time benchmark</summary>

```
Warming up --------------------------------------
       each - size 0     2.342M i/100ms
       to_h - size 0     2.695M i/100ms
Calculating -------------------------------------
       each - size 0     23.373M (± 1.8%) i/s -    234.222M in  10.024508s
       to_h - size 0     26.830M (± 4.5%) i/s -    269.514M in  10.080374s

Comparison:
       to_h - size 0: 26830441.8 i/s
       each - size 0: 23373275.0 i/s - 1.15x  slower

ruby 3.3.1 (2024-04-23 revision c56cd86388) [arm64-darwin23]
Warming up --------------------------------------
       each - size 1     2.342M i/100ms
       to_h - size 1     2.671M i/100ms
Calculating -------------------------------------
       each - size 1     23.174M (± 3.8%) i/s -    231.853M in  10.022762s
       to_h - size 1     26.821M (± 2.0%) i/s -    269.724M in  10.060542s

Comparison:
       to_h - size 1: 26820927.4 i/s
       each - size 1: 23174035.2 i/s - 1.16x  slower

ruby 3.3.1 (2024-04-23 revision c56cd86388) [arm64-darwin23]
Warming up --------------------------------------
       each - size 2     2.320M i/100ms
       to_h - size 2     2.697M i/100ms
Calculating -------------------------------------
       each - size 2     23.120M (± 4.4%) i/s -    232.039M in  10.059906s
       to_h - size 2     26.636M (± 3.8%) i/s -    267.023M in  10.041720s

Comparison:
       to_h - size 2: 26636324.0 i/s
       each - size 2: 23119650.1 i/s - 1.15x  slower

ruby 3.3.1 (2024-04-23 revision c56cd86388) [arm64-darwin23]
Warming up --------------------------------------
       each - size 5     2.245M i/100ms
       to_h - size 5     2.679M i/100ms
Calculating -------------------------------------
       each - size 5     23.150M (± 2.2%) i/s -    233.506M in  10.091786s
       to_h - size 5     26.758M (± 2.2%) i/s -    267.940M in  10.018463s

Comparison:
       to_h - size 5: 26758472.9 i/s
       each - size 5: 23149579.0 i/s - 1.16x  slower

ruby 3.3.1 (2024-04-23 revision c56cd86388) [arm64-darwin23]
Warming up --------------------------------------
      each - size 10     2.337M i/100ms
      to_h - size 10     2.683M i/100ms
Calculating -------------------------------------
      each - size 10     23.195M (± 2.1%) i/s -    233.656M in  10.078106s
      to_h - size 10     26.750M (± 4.2%) i/s -    268.296M in  10.052466s

Comparison:
      to_h - size 10: 26749930.0 i/s
      each - size 10: 23194783.9 i/s - 1.15x  slower

ruby 3.3.1 (2024-04-23 revision c56cd86388) [arm64-darwin23]
Warming up --------------------------------------
     each - size 100     2.311M i/100ms
     to_h - size 100     2.702M i/100ms
Calculating -------------------------------------
     each - size 100     23.238M (± 2.6%) i/s -    233.421M in  10.051715s
     to_h - size 100     26.877M (± 1.6%) i/s -    270.231M in  10.057313s

Comparison:
     to_h - size 100: 26876515.7 i/s
     each - size 100: 23238401.5 i/s - 1.16x  slower

ruby 3.3.1 (2024-04-23 revision c56cd86388) [arm64-darwin23]
Warming up --------------------------------------
    each - size 1000     2.326M i/100ms
    to_h - size 1000     2.678M i/100ms
Calculating -------------------------------------
    each - size 1000     23.339M (± 1.3%) i/s -    234.906M in  10.066823s
    to_h - size 1000     26.900M (± 1.5%) i/s -    270.489M in  10.057765s

Comparison:
    to_h - size 1000: 26900312.1 i/s
    each - size 1000: 23338519.6 i/s - 1.15x  slower
```

```ruby
require 'benchmark/ips'
#require 'active_support/core_ext/object/json'
require 'active_support/all'

H = {}

[0, 1, 2, 5, 10, 100, 1000].each do |hash_size|
  H.clear
  H.merge((0...hash_size).to_h { |i| ["key_#{i}", i] })

  Benchmark.ips do |x|
    x.config(warmup: 3, time: 10)
  
    x.report("each - size #{hash_size}") do |times|
      i = 0
      while (i += 1) <= times
        result = {}
        H.each { |k, v| result[k.to_s] = v.as_json }
      end
    end
    
    x.report("to_h - size #{hash_size}") do |times|
      i = 0
      while (i += 1) <= times
        H.to_h { |k, v| [k.to_s, v.as_json] }
      end
    end
    
    x.compare!
  end
end
```

</details>

But it allocates more intermediate garbage (1 per hash k/v pair, for that returned array):

<details><summary>Memory benchmark</summary>

Results:

```
Calculating -------------------------------------
                each   544.000  memsize (     0.000  retained)
                         1.000  objects (     0.000  retained)
                         0.000  strings (     0.000  retained)
                to_h   864.000  memsize (     0.000  retained)
                        11.000  objects (     0.000  retained)
                         0.000  strings (     0.000  retained)

Comparison:
                each:        544 allocated
                to_h:        864 allocated - 1.59x more
```

```ruby
#!/usr/bin/ruby

require "benchmark/memory"
require 'active_support/all'

H = (0...10).to_h { |i| ["key_#{i}", i] }

Benchmark.memory do |x|
  x.report("each") do
    result = {}
    H.each { |k, v| result[k.to_s] = v.as_json }
  end
  
  x.report("to_h") do
    H.to_h { |k, v| [k.to_s, v.as_json] }
  end
  
  x.compare!
end
```

</details>

WDYT?

Reply:
I think I'd keep as is. memory pressure can trigger more GC